zio-cache/shared/src/main/scala/zio/cache/Cache.scala

```diff
@@ -137,50 +145,80 @@ object Cache {
+              val builder          = ChunkBuilder.make[MapValue.Pending[Key, Error, Value]]()
+              val iterator         = keys.iterator
+              val size             = keys.size
+              val values           = Array.ofDim[MapValue[Key, Error, Value]](size)
+              var i                = 0
+              while (iterator.hasNext) {
+                val key   = iterator.next()
+                val value = map.get(key)
+                if (value eq null) {
+                  val promise = Promise.unsafeMake[Error, Value](fiberId)
+                  val pending = MapValue.Pending(new MapKey(key), promise)
+                  val value   = map.putIfAbsent(key, pending)
+                  if (value eq null) {
+                    trackMiss()
+                    builder += pending
+                    values(i) = pending
+                  } else {
+                    trackHit()
+                    values(i) = value
+                  }
+                } else {
+                  trackHit()
+                  values(i) = value
+                }
+                i += 1
               }
-              if (value eq null) {
-                trackAccess(key)
-                trackMiss()
-                lookup(k)
-                  .provide(environment)
-                  .run
-                  .flatMap { exit =>
-                    val now        = Instant.now()
-                    val entryStats = EntryStats(now)
-
-                    map.put(k, MapValue.Complete(key, exit, entryStats))
-                    promise.done(exit) *> ZIO.done(exit)
+              val pending          = builder.result()
+              pending.foreach(pending => trackAccess(pending.key))
+              val completePromises = lookup(pending.map(_.key.value))
+                .provide(environment)
+                .run
+                .flatMap { exit =>
+                  val now        = Instant.now()
+                  val entryStats = EntryStats(now)
+
+                  exit match {
+                    case Exit.Success(values) =>
+                      ZIO.foreach_(pending.zip(values)) { case (pending, value) =>
+                        map.put(pending.key.value, MapValue.Complete(pending.key, Exit.succeed(value), entryStats))
+                        pending.promise.succeed(value)
+                      }
+
+                    case Exit.Failure(cause) =>
+                      ZIO.foreach_(pending) { pending =>
+                        map.put(pending.key.value, MapValue.Complete(pending.key, Exit.halt(cause), entryStats))
+                        pending.promise.halt(cause)
+                      }
                   }
-                  .onInterrupt(promise.interrupt *> ZIO.succeed(map.remove(k)))
-              } else {
+                }
+                .onInterrupt(
+                  ZIO.collectAll(pending.map(_.promise.interrupt)) *>
+                    ZIO.succeed(pending.foreach(key => map.remove(key.key)))
```

Comment:
Without loading this PR into IntelliJ, I wonder if `map` here has a `removeAll` method, which can simplify things here.

Reply:
For Java's HashMap, we can do the following to remove a set of keys but it requires a Java HashSet so we'd need to use `CollectionConverters`/`JavaConverters` - probably not worth the trouble. :)
```scala
ZIO.succeed(
  map.keySet.removeAll(pending.map(_.key).toSet)
).ignore.unit
```